### ISSUE TITLE
Ambiguous dependency problem

### ISSUE THREAD

**Issue by ulrichSchreiner** [caddyserver/caddy#4331]
hi,

when i try to compile the 2.4.5 version i get the following error:
~~~
        github.com/caddyserver/caddy/v2/modules/caddyhttp imports
        github.com/google/cel-go/cel imports
        github.com/google/cel-go/parser imports
        github.com/antlr/antlr4/runtime/Go/antlr loaded from github.com/antlr/antlr4/runtime/Go/antlr@v0.0.0-20210826220005-b48c857c3a0e,
        but go 1.16 would fail to locate it:
        ambiguous import: found package github.com/antlr/antlr4/runtime/Go/antlr in multiple modules:
        github.com/antlr/antlr4 v0.0.0-20200503195918-621b933c7a7f (/home/usc/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200503195918-621b933c7a7f/runtime/Go/antlr)
        github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20210826220005-b48c857c3a0e (/home/usc/go/pkg/mod/github.com/antlr/antlr4/runtime/!go/antlr@v0.0.0-20210826220005-b48c857c3a0e)
~~~

this is NOT a problem of your code, it is a bug in `cel-go` which is already fixed (https://github.com/google/cel-go/issues/441). But as you have the v0.7.3 as a dependency and there is (at the moment) no new release, i had to patch my `go.mod` file to replace the library:
~~~
replace github.com/google/cel-go v0.7.3 => github.com/google/cel-go v0.7.4-0.20210818044834-cd8abd3c0e5e
~~~
I'm not sure if it a good idea to update your dependency too to this this commit-version?

It is ok to simply close this issue, as your code has no direct defect (in this case). this is only for documentation when other people have the same problem when they try to compile from a completely fresh codebase/workspace.

**Comment by francislavoie** [comment]
I'm confused. Why does this cause a build issue for you, but not for us in CI or releases? How are you building Caddy, exactly?

**Comment by ulrichSchreiner** [comment]
hi,
i'm building my custom service and everything works, until i do a `go get -u` to update my dependencies. so for example:
~~~go
package main

import (
        "net/http"
        "github.com/caddyserver/caddy/v2"
        "github.com/caddyserver/caddy/v2/modules/caddyhttp"
)

var (
        _ caddyhttp.MiddlewareHandler = (*MyApp)(nil)
)

type MyApp struct {
}

func (MyApp) CaddyModule() caddy.ModuleInfo {
        return caddy.ModuleInfo{
                ID:  "myapp",
                New: func() caddy.Module { return new(MyApp) },
        }
}

func init() {
        caddy.RegisterModule(MyApp{})
}

func (m *MyApp) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
        return nil
}

func main () {
}
~~~
this code compiles when i use v2.4.5 of caddy. when i now do a `go get -u` go should update the dependencies according to go's resolving algorithm (it should update to newer versions of compatible libraries).
~~~
❯ go get -u
mycaddy imports
        github.com/caddyserver/caddy/v2/modules/caddyhttp imports
        github.com/google/cel-go/cel imports
        github.com/google/cel-go/parser imports
        github.com/antlr/antlr4/runtime/Go/antlr: ambiguous import: found package github.com/antlr/antlr4/runtime/Go/antlr in multiple modules:
        github.com/antlr/antlr4 v0.0.0-20200503195918-621b933c7a7f (/home/usc/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200503195918-621b933c7a7f/runtime/Go/antlr)
        github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20210826220005-b48c857c3a0e (/home/usc/go/pkg/mod/github.com/antlr/antlr4/runtime/!go/antlr@v0.0.0-20210826220005-b48c857c3a0e)
~~~
in this small example it is not a real problem because the code is so functionless, that there is no need to update the libraries. but my own code has much more libraries inside and i really like to use `go get -u` to regularly update to the latest versions of the libraries to get patches, fixes, ... 
the only workaround was to do a replacement in `go.mod`  (or update every depencency separately, which is a mess).

as i wrote in the issue, this is not a problem of your code! but when using v2.4.5 as a library it is very likely to stumble over this issue in `cel-go`. so don't take this issue too seriously, i only wanted other people to have a chance to find this issue if they stumble over the same problem (my first action was to search caddy's issuetracker, so i think other would do the same).

**Comment by mholt** [comment]
Ah... yeah I don't believe using `go get -u` is a good idea (IMO) unless you know you want to upgrade the dependencies.

This is the main reason we're not going to be updating dependencies as often (unless there's a good reason to) -- if it ain't broke, don't fix it, kinda thing.

Thanks for documenting, but closing since I don't think there's anything for us to do here.

**Comment by mholt** [comment]
Interestingly, I am now encountering this error, even though I'm not using `go get -u`, nor have I changed any cel-go version. :thinking:

**Comment by mholt** [comment]
Ok, well, somehow these line got added to go.mod locally:

```
	github.com/antlr/antlr4 v0.0.0-20200503195918-621b933c7a7f // indirect
	github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20210826220005-b48c857c3a0e // indirect
```

and I just had to remove them.

**Comment by mohammed90** [comment]
> I'm having this happen with just this:
> 
> ```
> xcaddy build http-redirect-wrapper-module --with github.com/mastercactapus/caddy2-proxyprotocol
> ```
> 
> 😩
> 
> /cc @mohammed90

This is the culprit commit: https://github.com/caddyserver/caddy/commit/ff74a0aa09a24afd63686c2d79cb605ecbb47b8c. The build is ok before it but breaks from that commit on.

**Comment by mholt** [comment]
That's a bummer, because it was apparently fixed in v0.8.0.

Workaround was merged though. :(

**Comment by TristonianJones** [comment]
I've tried running the `xcaddy` tool against the `cel-upgrade` with a pre-release version of `cel-go@v0.11.0-pre` which vendors dependencies and creates a separate submodule for the conformance related `cel-go/server` subpackage, but this doesn't seem to help.

However, I did get the build to succeed with the following replacement:

```
xcaddy build cel-upgrade --with github.com/mastercactapus/caddy2-proxyprotocol --with github.com/antlr/antlr4=github.com/antlr/antlr4/runtime/Go/antlr@v0.0.0-20220209173558-ad29539cd2e9
```

Basically, I made sure that the old and new ANTLR versions are pointing at the same place. Perhaps, xcaddy could be patched with an option to automatically replace the antlr depedendency?

**Comment by francislavoie** [comment]
I updated the `cel-upgrade` branch to point to use `v0.11.0-pre`, and yeah, still not working:

```
$ xcaddy build cel-upgrade --with github.com/mastercactapus/caddy2-proxyprotocol
...
caddy imports
	github.com/caddyserver/caddy/v2/modules/standard imports
	github.com/caddyserver/caddy/v2/modules/caddyhttp/standard imports
	github.com/caddyserver/caddy/v2/modules/caddyhttp imports
	github.com/google/cel-go/cel imports
	github.com/google/cel-go/parser imports
	github.com/antlr/antlr4/runtime/Go/antlr: ambiguous import: found package github.com/antlr/antlr4/runtime/Go/antlr in multiple modules:
	github.com/antlr/antlr4 v0.0.0-20200503195918-621b933c7a7f (/home/francis/go/pkg/mod/github.com/antlr/antlr4@v0.0.0-20200503195918-621b933c7a7f/runtime/Go/antlr)
	github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20220209173558-ad29539cd2e9 (/home/francis/go/pkg/mod/github.com/antlr/antlr4/runtime/!go/antlr@v0.0.0-20220209173558-ad29539cd2e9)
```

Trying your suggestion with the antlr replacement, fails for me:

```
$ xcaddy build cel-upgrade --with github.com/mastercactapus/caddy2-proxyprotocol --with github.com/antlr/antlr4=github.com/antlr/antlr4/runtime/Go/antlr@v0.0.0-20220209173558-ad29539cd2e9
...
2022/03/15 02:22:54 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v github.com/mastercactapus/caddy2-proxyprotocol 
go: github.com/antlr/antlr4/runtime/Go/antlr@v0.0.0-20220209173558-ad29539cd2e9 used for two different module paths (github.com/antlr/antlr4 and github.com/antlr/antlr4/runtime/Go/antlr)
```

**Comment by TristonianJones** [comment]
Ah, I think I see what the difference is now. Two quick things:

1. The replace with my private copy of cel-go is still present in the `go.mod` in addition to the new pre-release
2. The antlr version cannot be the same as the version used by a cel-go release. Since these are only pseudo-versions we just need to pick a version CEL has never used to make the upgrade work:

```
xcaddy build cel-upgrade \
    --with github.com/mastercactapus/caddy2-proxyprotocol \
    --with github.com/antlr/antlr4=github.com/antlr/antlr4/runtime/Go/antlr@v0.0.0-20220314183648-97c793e446ba
```

In the command line you ran earlier,  `v0.0.0-20220209173558-ad29539cd2e9` is the pseudo-version of ANTLR referenced in the cel-go v0.11.0-pre release, which appears to make the build complain. In my local tests, I was still running against the cel-go v0.9.0 release which refers to an older pseudo-version of ANTLR.

Give it another try with the recommended changes. I think you'll see a positive result.

**Comment by francislavoie** [comment]
Yeah, that last one does work.

I still really don't like this though. Seems like such a terrible and brittle hack :weary:

**Comment by TristonianJones** [comment]
@mohammed90 when you mention "why it was added in the first place" are you talking about the `go.mod` to ANTLR or something else?

**Comment by mohammed90** [comment]
> @mohammed90 when you mention "why it was added in the first place" are you talking about the `go.mod` to ANTLR or something else?

No, rather why `github.com/antlr/antlr4` is added to our build despite caddy depending on newer version of cel-go, which does not depend on `github.com/antlr/antlr4` rather depends on `github.com/antlr/antlr4/runtime/Go/antlr`. I'd expect only `github.com/antlr/antlr4/runtime/Go/antlr` is added to our build because our dependency graph points at cel-go version that depends on it, not `github.com/antlr/antlr4`.

Having the module `github.com/mastercactapus/caddy2-proxyprotocol` depend on an older version of caddy shouldn't pull `github.com/antlr/antlr4` because the caddy version we're depending on depends on new cel-go module that doesn't depend on `github.com/antlr/antlr4`.

I hope that clears out the logic :joy: I feel like I need to graph the thing to explain it

**Comment by francislavoie** [comment]
FYI @TristonianJones see https://github.com/caddyserver/xcaddy/pull/92, we apparently have a really weird fix which just involves running `go get -d -v` which triggers downloading the right `antlr4` version and solves the ambiguity.

**Comment by TristonianJones** [comment]
@francislavoie wow, that's a really happy accident. Does this mean it's worth looking at upgrading CEL again and reviving the `HttpRequest` matchers investigation?

**Comment by francislavoie** [comment]
Yeah, we'll probably upgrade CEL soon, after the fix in `xcaddy` is merged and released.

**Comment by TristonianJones** [comment]
Awesome. Then I'll give that other issue a look to see what I can come up with there. The v0.11.2 vastly improves CEL environment bootstrapping costs and has some evaluation performance improvements as well.
